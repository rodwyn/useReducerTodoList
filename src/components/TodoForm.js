import React from 'react';
import { useForm } from "../hooks/useForm";

export const TodoForm = ({handleAddTodo}) => {
  const [{description}, handleInputChange, reset] = useForm({
    description: ''
  });

  const handleSubmit = event => {
    event.preventDefault();
    
    if (description.trim().length < 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    };

    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Add ToDo</h4>
      <hr />
      <form onSubmit={ handleSubmit }>
        <input
          type="text"
          name="description"
          placeholder="Add Todo"
          autoComplete="off"
          className="form-control"
          onChange={handleInputChange}
          value={description}
            />
        <div className="d-grid gap-2 mt-2">
          <button className="btn btn-outline-primary" type="submit">Add</button>
        </div>
      </form>
    </>
  )
}
