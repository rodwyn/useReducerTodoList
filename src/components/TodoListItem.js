import React from 'react'

export const TodoListItem = ({todo, i, handleToggle, handleDelete}) => {
  return (
    <li
    className="list-group-item"
    key={todo.id}>
      <p
        className={`${todo.done && 'completed'}`}
        onClick={() => handleToggle(todo.id)}>{i + 1}.{todo.desc}</p>
      <button className="btn btn-danger" onClick={() => handleDelete(todo.id)}>Delete</button>
    </li> 
  )
}
