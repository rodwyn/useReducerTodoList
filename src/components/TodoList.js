import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, handleToggle, handleDelete}) => {
  return (
    <ul className="list-group list-group-flush">
      {
        todos.map((todo, i) => (
          <TodoListItem
            todo={todo}
            handleToggle={handleToggle}
            handleDelete={handleDelete} 
            i={i} 
            key={todo.id}
            />
        ))
      }
    </ul>
  )
}
