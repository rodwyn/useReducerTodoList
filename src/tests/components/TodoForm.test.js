import { shallow } from 'enzyme';
import React from 'react';
import { TodoForm } from '../../components/TodoForm';

describe('Testing <TodoForm /> component.', () => {
  const handleAddTodo = jest.fn();
  const wrapper = shallow(<TodoForm handleAddTodo={handleAddTodo} />);

  test('should render correctly.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should not call the handleAddTodo function.', () => {
    const formSubmit = wrapper.find('form').prop('onSubmit');
    
    formSubmit({preventDefault(){}});
    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });
  
  test('should call handleAddTodo on form interaction.', () => {
    const value = 'Learn to fly.';
    wrapper.find('input').simulate('change', {
      target: {
        value,
        name: 'description'
      }
    });
    const formSubmit = wrapper.find('form').prop('onSubmit');
    
    formSubmit({preventDefault(){}});
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      desc: value,
      done: false
    });
    expect(wrapper.find('input').prop('value')).toBe('');
  })
  
});