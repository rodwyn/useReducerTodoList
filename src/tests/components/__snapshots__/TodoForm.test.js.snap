// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <TodoForm /> component. should render correctly. 1`] = `
<Fragment>
  <h4>
    Add ToDo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Add Todo"
      type="text"
      value=""
    />
    <div
      className="d-grid gap-2 mt-2"
    >
      <button
        className="btn btn-outline-primary"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
</Fragment>
`;
